### ISSUE TITLE
Fatal error on first migration run MySQL in Ubuntu

### ISSUE THREAD

**Issue by MattMorgis** [db-migrate/node-db-migrate#311]
I am working on an Ubuntu 14.04 server with MySQL 5.6. I am running nom v2.14.2 and node v4.0.0.

I checked out my repo and on my first attempt to setup the database I keep getting fatal error. Here is the verbose output:

```
db-migrate up -v
[INFO] Using dev settings: { host: '127.0.0.1',
  user: 'root',
  password: '***************************',
  database: '<databases-name>',
  driver: 'mysql',
  multipleStatements: true }
[INFO] require: ./mysql
[INFO] connecting
[INFO] connected
[INFO] creating table: migrations
[SQL] CREATE TABLE IF NOT EXISTS `migrations` (`id` INTEGER  PRIMARY KEY AUTO_INCREMENT NOT NULL, `name` VARCHAR (255) NOT NULL, `run_on` DATETIME  NOT NULL)
[INFO] migration table created
[INFO] loading migrations from dir /home/<user>/<project>/db/migrations
[INFO] loading migrations from database
[SQL] SELECT * FROM `migrations` ORDER BY run_on DESC, name DESC
[ERROR] Error: Cannot enqueue Query after fatal error.
    at Protocol._validateEnqueue (/usr/local/lib/node_modules/db-migrate/node_modules/mysql/lib/protocol/Protocol.js:193:16)
    at Protocol._enqueue (/usr/local/lib/node_modules/db-migrate/node_modules/mysql/lib/protocol/Protocol.js:129:13)
    at Connection.query (/usr/local/lib/node_modules/db-migrate/node_modules/mysql/lib/Connection.js:201:25)
    at Class.Base.extend.all (/usr/local/lib/node_modules/db-migrate/lib/driver/mysql.js:414:34)
    at Class.Base.extend.allLoadedMigrations (/usr/local/lib/node_modules/db-migrate/lib/driver/mysql.js:424:10)
    at Function.Migration.loadFromDatabase (/usr/local/lib/node_modules/db-migrate/lib/migration.js:217:10)
    at /usr/local/lib/node_modules/db-migrate/lib/migrator.js:103:17
    at /usr/local/lib/node_modules/db-migrate/lib/migration.js:211:5
    at FSReqWrap.oncomplete (fs.js:82:15)
```
I am not sure if this is a problem with my setup or db-migrate. I can run the migrations on my OS X local machine just fine. 

I also can log into the MySQL console on the Ubuntu server and create the migrations table fine, making me think it may not be a permission error either.

```
mysql>CREATE TABLE IF NOT EXISTS `migrations` (`id` INTEGER  PRIMARY KEY AUTO_INCREMENT NOT NULL, `name` VARCHAR (255) NOT NULL, `run_on` DATETIME  NOT NULL);
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT * FROM `migrations` ORDER BY run_on DESC, name DESC;
Empty set (0.00 sec)

mysql> DROP TABLE migrations;
```

**Comment by MattMorgis** [comment]
My first migration is as follows:

```
BEGIN;

CREATE USER 'appdev'@'localhost';
GRANT ALL PRIVILEGES ON `database\_v1`.* TO 'appdev'@'localhost' WITH GRANT OPTION;

CREATE USER 'appdbuser'@'%.domain.com';
GRANT ALL PRIVILEGES ON `database\_v1`.* TO 'appdbuser'@'%.domainname.com' REQUIRE SSL WITH GRANT OPTION;

FLUSH PRIVILEGES;

COMMIT;
```

**Comment by wzrdtales** [comment]
Is this really the value of the database name in your configfile?

    database: '<databases-name>',

**Comment by wzrdtales** [comment]
If yes, this could be the failure. And it seems to be that you specified invalid values in your configuration as when loading the data sets from your migrations table: at Class.Base.extend.allLoadedMigrations (/usr/local/lib/node_modules/db-migrate/lib/driver/mysql.js:424:10)

**Comment by MattMorgis** [comment]
Um.. no. I changed for the sole purpose of posting the ticket. Again, I check the repo out on my OS X machine and run `db-migrate up` and it creates the migrations table and runs through the commands fine. I do it on Ubuntu and I get that stack trace.

**Comment by MattMorgis** [comment]
Mac OS X stack trace with empty database and no migrations table. 
```
db-migrate up -v
[INFO] Using dev settings: { host: '127.0.0.1',
  user: 'root',
  password: '******',
  database: '<databasename>',
  driver: 'mysql',
  multipleStatements: true }
[INFO] require: ./mysql
[INFO] connecting
[INFO] connected
[INFO] creating table: migrations
[SQL] CREATE TABLE IF NOT EXISTS `migrations` (`id` INTEGER  PRIMARY KEY AUTO_INCREMENT NOT NULL, `name` VARCHAR (255) NOT NULL, `run_on` DATETIME  NOT NULL)
[INFO] migration table created
[INFO] loading migrations from dir /Users/mattmorgis/Development/NodeJS/<project>/db/migrations
[INFO] loading migrations from database
[SQL] SELECT * FROM `migrations` ORDER BY run_on DESC, name DESC
[INFO] preparing to run up migration: 20150919152919-add-app-users
```

**Comment by wzrdtales** [comment]
Yes and this stack trace indicates an error while loading the migrations from your table. Please make sure your configuration and permission fits.

I just tested on the exact same environment you described and had no issues, so this is definitely a configuration issue.

**Comment by wzrdtales** [comment]
Maybe you can post your config files? Naturally star out the password and secret stuff ;)

**Comment by MattMorgis** [comment]
```{
    "dev": {
	    "host": "127.0.0.1",
	    "user": { "ENV" : "DB_USER" },
	    "password" : { "ENV" : "DB_PASS" },
	    "database": "database_v1",
	    "driver": "mysql",
	    "multipleStatements": true
  	},
  	"sql-file" : true
}```

**Comment by MattMorgis** [comment]
Why would the same config file work on OS X and not linux? And why can I log into the console on my Ubuntu server and run the same commands without any permissions errors? I can also echo DB_USER and DB_PASS. db-migrate is obviously picking them up because it prints "root" in the stack trace.

**Comment by MattMorgis** [comment]
How can I add additional logging to get a more descriptive error message?

**Comment by wzrdtales** [comment]
Ok I tested with the exact same configuration. And I was able to reproduce the same stack trace you've got. My issue was, that the database hasn't exist. After I created it, everything works again. Maybe you've got the same issue?

I will see to make the logging include more informations in case of errors.

**Comment by MattMorgis** [comment]
No. See screen shot. The database is created and correctly named. I've checked that about 15 times. 
![screen shot 2015-09-21 at 8 40 10 am](https://cloud.githubusercontent.com/assets/4664374/9991951/8e511e2e-603c-11e5-87bf-fc0e02bc7275.png)

**Comment by wzrdtales** [comment]
Then the database user can not see this database for some reason, probably a permission problem. You know that the permissions for the user root@127.0.0.1 are not automatically the same as for root@localhost?

**Comment by wzrdtales** [comment]
#312 created and planned to make such tasks easier in the future.

**Comment by MattMorgis** [comment]
I had localhost to start and changed to 127.0.0.1 to see if that made a difference. Again running `mysql -u root -p` and then creating the table myself works fine. If the database didn't exist or it was a permission error, wouldn't the migrations table not get created in the first place the program bomb out there?

```
mysql>CREATE TABLE IF NOT EXISTS `migrations` (`id` INTEGER  PRIMARY KEY AUTO_INCREMENT NOT NULL, `name` VARCHAR (255) NOT NULL, `run_on` DATETIME  NOT NULL);
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT * FROM `migrations` ORDER BY run_on DESC, name DESC;
Empty set (0.00 sec)

mysql> DROP TABLE migrations;
```

**Comment by wzrdtales** [comment]
For some strange reason it doesn't fail while creating the migratin table. I have seen the same behavior on my side. I'm going to look after this how ever.

If you do a `mysql -u root -p` you will always connect via the unix socket and thus connect as root@localhost not as root@127.0.0.1

**Comment by MattMorgis** [comment]
Okay. Changing 127.0.0.1 to localhost in my config file gives me the same results.

**Comment by wzrdtales** [comment]
Root should normally be able to see all dbs, this is strange enough that there seems to be a problem with this. I'm unfortunately not able to reproduce your issue with the current amount of informations. Could you maybe prepare a gist with an example that fails at your side and try to include the permissions you set and so on?

Or you could create a mini script and run queries with your config via node and check the results?

**Comment by MattMorgis** [comment]
Is there anywhere I can add a log statement to get the exact error MySQL is returning? That would probably help narrow this down more. I'm not sure how much more information I can give you. I created the server. I ran
 `apt-get install mysql-server` 
`mysql_secure_installation`
`mysql -u root -p`
`mysql> CREATE DATABASE databasename_v1;`
`apt-get install git`
`git clone https://github.com/repo.git`
`npm install -g db-migrate`
`db-migrate up`

**Comment by MattMorgis** [comment]
In the my.cnf file I have the bind-address property set to the external IP of the server. Could that have anything to do with this? I will try running on a fresh install again. But getting the exact MySQL error would make this less of a wild goose chase.

**Comment by wzrdtales** [comment]
You would need to get in the runsql function and add a console.log in the callback of it.

I have tested some other things right now. And I saw, that if the password is wrong I also get the same stack trace. Maybe this is your issue?

**Comment by wzrdtales** [comment]
The error handling is definitely a topic I wanted to rework, but there are some things discovered by this issue I see that are missing now. The mysql driver implicitly opens the connection on the first query, and thus this case isn't handled.

I'm away for some hours now, please check if the password in the environment is correct. May be it has an empty space at the end or something like this. I'm coming back to you as soon as I'm available again.

**Comment by MattMorgis** [comment]
I'm in the same boat and have to step away for a bit too. I'll double check the password and let you know. I'll also add in that console.log when I get a chance later and post the response.  Thank you for the help.

**Comment by MattMorgis** [comment]
After some additional logging, I found this `{ [Error: Cannot enqueue Query after fatal error.] code: 'PROTOCOL_ENQUEUE_AFTER_FATAL_ERROR', fatal: false }` which seems to be related to the node-mysql library. 

https://github.com/felixge/node-mysql/issues?q=is%3Aissue+%7B+%5BError%3A+Cannot+enqueue+Query+after+fatal+error.%5D+code%3A+%27PROTOCOL_ENQUEUE_AFTER_FATAL_ERROR%27%2C+fatal%3A+false+%7D+is%3Aclosed

Are you using pools to connect to the database?

**Comment by wzrdtales** [comment]
No we don't use pools and this is also controlled by the configuration you provide though. However, the problem right now is, if the node driver is running into an error, the error doesn't seem to be thrown and the **next** execution of some command is going to fail. This means, actually not the latest executed command is the problem, but the one before it.

I'm sorry that I'm away again and first available in the evening again (that means in about 5 hours from now on). I will look to prepare a branch for you which is throwing more errors, maybe you have found the solution by then already, but I will rework the error reporting anyway, though this is a good preparation step for the real rework.

**Comment by MattMorgis** [comment]
I have figured out the cause of the issue. I have edited the MySQL configuration file (/etc/mysql/my.cnf) and changed the `bind-address` property from 127.0.0.1 to the IP of the server. This is acting as a remote DB instance. When I changed the IP back to 127.0.0.1, `db-migrate up` worked just fine. Changing it back to the IP address caused the crash again. 

I followed these instructions: https://www.digitalocean.com/community/tutorials/how-to-set-up-a-remote-database-to-optimize-site-performance-with-mysql

**Comment by MattMorgis** [comment]
Now we have a starting point for when you are available again. I need to accept external database connections and not just ones from localhost.

**Comment by wzrdtales** [comment]
Ok, this is actually not a failure of db-migrate, but of your configuratoin.

If you're configuring your MySQL/MariaDB server to listen on the external ip, you need to specify also the external ip in your db-migrate configuartion. Alternatively you can do two things.

1. use unix sockets, look up the configs for this here https://github.com/felixge/node-mysql/#connection-options
2. bind your MySQL/MariaDB server to 0.0.0.0 instead, this actually depends on your needs.

**Comment by wzrdtales** [comment]
@MattMorgis I've just published v0.9.23 this version has got a patch which throws now an exception if the connection establishment has not worked properly.

**Comment by MattMorgis** [comment]
So after messing with the user's and bind-address I think I've got this all sorted out. I don't believe node-mysql is connecting via localhost so I needed to setup a user like `'username'@'privatenetworkIP' instead of 'user'@'localhost' in the config. I also noticed that the patch now includes the SQL error message. This issue can now be closed.
